### ISSUE TITLE
[MSHADE-171] "Access Denied" on Windows because shade plugin seems to try to open a directory as Zip-File

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#530]
**[Dominik Stadler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dominik.stadler@gmx.at)** opened **[MSHADE-171](https://issues.apache.org/jira/browse/MSHADE-171?redirect=false)** and commented

Config of plugin is fairly simple, but fails with the error below when I call "mvn package test", it works when I call only "mvn package", it seems the plugin tries to open a directory as zip-file.

It happens in a multi-module setup when building from the parent-directory. It seems the dependency between the module where the shade plugin runs onto another module causes the target/classes of that module to be added as artifact-file, which causes the shade plugin to fail.

Let me know if you need a self-contained reproducer project, I can provide one, but it will take some time to build it...

FYI, there was a similar problem in the war plugin at MWAR-192.

```
<plugin>
     <groupId>org.apache.maven.plugins</groupId>
     <artifactId>maven-shade-plugin</artifactId>
     <version>2.3</version>
     <executions>
          <execution>
                   <phase>package</phase>
                   <goals>
                            <goal>shade</goal>
                   </goals>
          </execution>
     </executions>
</plugin>
```

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:2.3:shade (default) on project towerrunner-ui: Error creating shaded jar: C:\workspaces\workspace.GWT.changeset\towerrunner\towerrunner-base\target\classes (Access is denied) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:2.3:shade (default) on project towerrunner-ui: Error creating shaded jar: C:\workspaces\workspace.GWT.changeset\towerrunner\towerrunner-base\target\classes (Access is denied)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:108)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:116)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:361)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error creating shaded jar: C:\workspaces\workspace.GWT.changeset\towerrunner\towerrunner-base\target\classes (Access is denied)
        at org.apache.maven.plugins.shade.mojo.ShadeMojo.execute(ShadeMojo.java:566)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 19 more
Caused by: java.io.FileNotFoundException: C:\workspaces\workspace.GWT.changeset\towerrunner\towerrunner-base\target\classes (Access is denied)
        at java.util.zip.ZipFile.open(Native Method)
        at java.util.zip.ZipFile.<init>(ZipFile.java:215)
        at java.util.zip.ZipFile.<init>(ZipFile.java:145)
        at java.util.jar.JarFile.<init>(JarFile.java:153)
        at java.util.jar.JarFile.<init>(JarFile.java:117)
        at org.apache.maven.plugins.shade.DefaultShader.newJarFile(DefaultShader.java:267)
        at org.apache.maven.plugins.shade.DefaultShader.shade(DefaultShader.java:100)
        at org.apache.maven.plugins.shade.mojo.ShadeMojo.execute(ShadeMojo.java:471)
        ... 21 more
[ERROR] 
```

---

**Affects:** 2.3

**Attachments:**
- [MSHADE-171.zip](https://issues.apache.org/jira/secure/attachment/12719019/MSHADE-171.zip) (_3.06 kB_)

**Issue Links:**
- [MSHADE-213](https://issues.apache.org/jira/browse/MSHADE-213) Clarify contract of ResourceTransformer wrt closing of streams

- [MSHADE-214](https://issues.apache.org/jira/browse/MSHADE-214) Update jdependency to 1.1 to fix resource leaks

1 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

It would be great if you could provide and example project which reproduce the behaviour. Thanks in advance.

**Comment by jira-importer** [comment]
**[Richard Sand](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rsand2)** commented

I have the same error - it occurs for me when one of the dependencies is found in the local file system as opposed to the local repository, for example in Eclipse when checking the box "resolve local workspace artifacts".

If the shade plugin did a simple "isDirectory" test before proceeding with unzipping/unjarring it would probably work fine in all circumstances.

-Richard

**Comment by jira-importer** [comment]
**[Dominik Stadler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dominik.stadler@gmx.at)** commented

Here is a test-project that has everything removed that is not related to triggering the problem, it should show the access denied error if you run the following on a Windows machine:

`mvn package test`

Running `package` or `test` standalone works fine.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

There was some logic that did not close streams properly upon errors in shade processing; this might be the cause of this problem. Fixed in r1713986, you may want to give 2.5-SNAPSHOT a try (I'm not on windows)

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Additional resource leaks have just been fixed in jdependency, so we'd need to wait for a release of a version >1.0
